### DISCUSSION TITLE
Chapter 9 - model.to(device)

### DISCUSSION THREAD

**Discussion by jeffsnell** [mrdbourke/pytorch-deep-learning#993] [Q&A]
Thank you for this class - I am learning a lot!!

I am running all your code on an Intel/Mac Pro (*mps* GPU) and everything runs well and nearly as fast as Colab in your video. I replace your device agnostic code block with the following:
```
# Setup device-agnostic code 
import torch
if torch.cuda.is_available():
    device = "cuda" # NVIDIA GPU
elif torch.backends.mps.is_available():
    device = "mps" # Apple GPU
else:
    device = "cpu" # Defaults to CPU if NVIDIA GPU/Apple GPU aren't available

# device = "cpu"
print(f"Using device: {device}")
```

I find that each time before calling the going_modular - **engine.train(...)** function - I need to add one line of code to move the **model.to(device)** or I get an error

I tried adding the **model.to(device)** code into the *engine.py* file - just before the **model.train()** call. But that does not seem to make a difference.

What am I missing here??

**Comment by pritesh2000** [comment]
`model.to(device)` used for sending model to target device. 
When you are using GPU/ CPU of any type sending model to target device means model will run on GPU/CPU whichever you chose and based on that your model will take time computing results.

ex. GPU will take less time for computing the same results than CPU

Here you might get error, because your training data and model are on different device (GPU/CPU) so you have put them on same device to get computation done.

**Reply by jeffsnell** [reply]
I understand the purpose of `model.to(device)`  - what I do not understand **why** when when I put that statement _inside_ the `engine.train(...)` function it does not seem to work while when II put that statement _before_ the `engine.train(...)` it does ??

Inside the `engine.train(...)` function I put `model.to(device)` just ahead of  the `model.train()` call. But it does not work there...

**Reply by pritesh2000** [reply]
Can you share your code ? (error)

If you modified train function then share that too
